### PR TITLE
added working code for BalanceStats and TextTable

### PR DESCRIPTION
**Pull request by ygnacio**
- Added code for methods in BalanceStats Class to calculate the highest balance, 
biggest loss, richest person and most poor person.

- Added code for the method Build in the class TextTable to build a message in a text table with given padding.

### CODE REVIEW

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 10
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  7
- Src/BootCamp.Chapter/Program.cs  1
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037&bid=16678002)

File: Src/BootCamp.Chapter/TextTable.cs [hunk 1]
```diff
@@ -34,7 +36,63 @@ public static class TextTable
...
+
+            // top border
+            string result = $"{corner}{topLine}{corner}{newLine}";
+            result += verticalPadding;
+
+            //rows
+            foreach (string item in stringArray)
+            {
+                string sentence = item.PadRight(maxLength);
+                result += $"{leftBorder}{sentence}{rightBorder}{newLine}";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 2]
```diff
@@ -7,31 +10,315 @@ public static class BalanceStats
...
+                if (min > currentBalance)
+                {
+                    min = currentBalance;
+                    poorPeople = auxBalances[0];
+                    result = $" has the least money. {min:C0}.";
+                    moreThanOne = false;
+                }
+                else if(min.Equals(currentBalance))
+                {
+                    poorPeople += $", {auxBalances[0]}";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 3]
```diff
@@ -7,31 +10,315 @@ public static class BalanceStats
...
+                if (max < currentBalance)
+                {
+                    max = currentBalance;
+                    richestPeople = auxBalances[0];
+                    moreThanOne = false;
+                    result = $" is the richest person. {max:C0}.";
+                }
+                else if(max.Equals(currentBalance))
+                {
+                    richestPeople += $", {auxBalances[0]}";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 4]
```diff
@@ -7,31 +10,315 @@ public static class BalanceStats
...
+                if (max < currentMax )
+                {
+                    max = currentMax;
+                    peopleWithSameHighBalance = auxBalances[0];
+                    moreThanOne = false;
+                    maxResult = $" had the most money ever. {max:C0}.";
+                }
+               else if (max.Equals(currentMax))
+                {
+                    peopleWithSameHighBalance += $", {auxBalances[0]}";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/TextTable.cs [hunk 5]
```diff
@@ -34,7 +36,63 @@ public static class TextTable
...
+            string leftBorder = "|" + horizontalPadding;
+            string rightBorder = horizontalPadding + "|";
+            string verticalPadding = "";
+            string newLine = Environment.NewLine;
+
+            if (padding > 0)
+            {
+                for (int i = 0; i < padding; i++)
+                {
+                    verticalPadding += leftBorder + new string(' ', maxLength)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/TextTable.cs [hunk 6]
```diff
@@ -34,7 +36,63 @@ public static class TextTable
...
+            foreach (string item in stringArray)
+            {
+                if (maxLength < item.Length)
+                {
+                    maxLength = item.Length;
+                }
+            }
+
+            // calculate table borders
+            char corner = '+';
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: Codacy Issue found: [Add the 'const' modifier to 'corner'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/Program.cs [hunk 7]
```diff
@@ -1,14 +1,27 @@
...
             // - FindHighestBalanceEver
+            stats[0] = BalanceStats.FindHighestBalanceEver(PeoplesBalances.Balances);
             // - FindPersonWithBiggestLoss
+            stats[1] = BalanceStats.FindPersonWithBiggestLoss(PeoplesBalances.Balances);
             // - FindRichestPerson
+            stats[2] = BalanceStats.FindRichestPerson(PeoplesBalances.Balances);
             // - FindMostPoorPerson
+            stats[3] = BalanceStats.FindMostPoorPerson(PeoplesBalances.Balances);
+
+            int padding = 3;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Codacy Issue found: [Add the 'const' modifier to 'padding'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 8]
```diff
@@ -7,31 +10,315 @@ public static class BalanceStats
...
+        }
+
+        /// <summary>
+        /// Replace the last comma with the string " and" in a string with values separated by comma.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns>Return a string with the last "," replaced by " and"</returns>
+        private static string ReplaceLastCommaWithAnd(string value) 
+        {
+            int lastComma = value.LastIndexOf(',');
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Remove this useless assignment to local variable 'lastComma'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 9]
```diff
@@ -7,31 +10,315 @@ public static class BalanceStats
...
+                if (min > currentMin)
+                {
+                    min = currentMin;
+                    peopleWithSameLowBalance = auxBalances[0];
+                    moreThanOne = false;
+                    minResult = $" lost the most money. {min:C0}.";
+                }
+                else if (min.Equals(currentMin))
+                {
+                    peopleWithSameLowBalance += $", {auxBalances[0]}";
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 10]
```diff
@@ -7,31 +10,315 @@ public static class BalanceStats
...
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private static bool IsDecArrayNullOrEmpty(decimal[] array)
+        {
+            if (array == null || array.Length == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 8 above)
Comment (ygnacio) [reply Src/BootCamp.Chapter/BalanceStats.cs]: I don't understand this Kai, I'm using the variable in the next sentence:
```
int lastComma = value.LastIndexOf(',');
return value[..lastComma] + " and" + value[(lastComma + 1)..];
```

Comment (Almantask) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Ignore it, it was a poor analysis of a bot.

Comment (ygnacio) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Although, I think I'll use the stringbuilder to do the same thing and delete the function, but still I'll use the variable to calculate the last comma position. I think it is more readable. For example:
```
StringBuilder b = new StringBuilder("uno, dos, tres y mas texto");
int pos = b.ToString().LastIndexOf(",");
b.Replace(",", " y", pos, 1); // b = "uno, dos y tres y mas texto"
```

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  7
- Src/BootCamp.Chapter/Program.cs  1
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037&bid=16678002)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 11]
```diff
@@ -7,31 +11,335 @@ public static class BalanceStats
...
+            return auxArray;
+        }
+
+        /// <summary>
+        /// Changes the Strinbuilder instance by Replacing the last comma with the string " and" in a string with values separated by comma.
+        /// </summary>
+        /// <param name="sb"></param>
+        private static void ReplaceLastCommaWithAnd(StringBuilder sb) 
+        {
+            int lastComma = sb.ToString().LastIndexOf(",");
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Define the locale to be used in this string operation.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 12]
```diff
@@ -7,31 +11,335 @@ public static class BalanceStats
...
+                {
+                    peopleWithSameLowBalance.Append($", {auxBalances[0]}");
+                    moreThanOne = true;
+                }
+            }
+
+            if (moreThanOne)
+            {
+                ReplaceLastCommaWithAnd(peopleWithSameLowBalance);
+                //int lastComma = peopleWithSameLowBalance.ToString().LastIndexOf(',');
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Remove this commented out code.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 13]
```diff
@@ -7,31 +11,335 @@ public static class BalanceStats
...
+            return (array == null || array.Length == 0) ? true : false;
+        }
+        /// <summary>
+        /// verify if decimal array is null or empty
+        /// </summary>
+        /// <param name="array"></param>
+        /// <returns></returns>
+        private static bool IsDecArrayNullOrEmpty(decimal[] array)
+        {
+            return (array == null || array.Length == 0) ? true : false;
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Codacy Issue found: [Remove the unnecessary Boolean literal(s).](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037)

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 11 above)
Comment (ygnacio) [reply Src/BootCamp.Chapter/BalanceStats.cs]: Ok, do I need to define locale in this case? I'm replacing the last comma in a list of names by an "and"

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  7
- Src/BootCamp.Chapter/Program.cs  1
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037&bid=16678002)

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 14]
```diff
@@ -7,31 +11,337 @@ public static class BalanceStats
...
+            return poorPeople.ToString();
+        }
+
+        /// <summary>
+        /// Makes a customizable invariant culture without currencygroupseparator and currencynegativepattern = 1
+        /// </summary>
+        /// <returns>an Invariant Culture CultureInfo</returns>
+        private static CultureInfo GetCustomInvariantCulture()
+        {
+            CultureInfo cult = (CultureInfo)CultureInfo.InvariantCulture.Clone();
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: Just call it full name- culture. cult gives it a whole different funny meaning.

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 15]
```diff
@@ -7,31 +11,337 @@ public static class BalanceStats
...
+
+            return cult;
+        }
+
+        /// <summary>
+        /// verify if string array is null or empty
+        /// </summary>
+        /// <param name="array"></param>
+        /// <returns></returns>
+        private static bool IsStrArrayNullOrEmpty(string[] array)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: You can just call it IsNullOrEmpty

File: Src/BootCamp.Chapter/BalanceStats.cs [hunk 16]
```diff
@@ -7,31 +11,337 @@ public static class BalanceStats
...
+        {
+            return (array == null || array.Length == 0);
+        }
+
+        /// <summary>
+        /// verify if decimal array is null or empty
+        /// </summary>
+        /// <param name="array"></param>
+        /// <returns></returns>
+        private static bool IsDecArrayNullOrEmpty(decimal[] array)
```
Comment (Almantask) [inline Src/BootCamp.Chapter/BalanceStats.cs]: You can just call it IsNullOrEmpty. Method overload will handle the rest.

File: Src/BootCamp.Chapter/Program.cs [hunk 17]
```diff
@@ -1,14 +1,27 @@
...
 {
     class Program
     {
         static void Main(string[] args)
         {
+            string[] stats = new string[4];
             // Print each of the statistical output using Text Table with padding 3:
             // - FindHighestBalanceEver
+            stats[0] = BalanceStats.FindHighestBalanceEver(PeoplesBalances.Balances);
             // - FindPersonWithBiggestLoss
```
Comment (Almantask) [inline Src/BootCamp.Chapter/Program.cs]: Comments are not needed. The code speaks for itself.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 18]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
...
-            return "";
+            if (string.IsNullOrEmpty(message))
+            {
+                return "";
+            }
+
+            // split the message
+            string[] stringArray = message.Split(Environment.NewLine);
+            int maxLength = -1;
+
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: One extra blank line.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 19]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
         /// </summary>
         public static string Build(string message, int padding)
         {
-            return "";
+            if (string.IsNullOrEmpty(message))
+            {
+                return "";
+            }
+
+            // split the message
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: move to a function instead of writing a comment.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 20]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
...
+            string[] stringArray = message.Split(Environment.NewLine);
+            int maxLength = -1;
+
+
+            // calculate max length
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: move to a function instead of writing a comment.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 21]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
...
+            foreach (string item in stringArray)
+            {
+                if (maxLength < item.Length)
+                {
+                    maxLength = item.Length;
+                }
+            }
+
+            // calculate table borders
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: move to a function instead of writing a comment.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 22]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
...
+
+            if (padding > 0)
+            {
+                for (int i = 0; i < padding; i++)
+                {
+                    verticalPadding.Append(leftBorder + new string(' ', maxLength) + rightBorder + newLine);
+                }
+            }
+
+            // top border
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: move to a function instead of writing a comment.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 23]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
...
+            StringBuilder result = new StringBuilder($"{corner}{topLine}{corner}{newLine}");
+            result.Append(verticalPadding);
+
+            //rows
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: move to a function instead of writing a comment.

File: Src/BootCamp.Chapter/TextTable.cs [hunk 24]
```diff
@@ -34,7 +37,62 @@ public static class TextTable
...
+            foreach (string item in stringArray)
+            {
+                string sentence = item.PadRight(maxLength);
+                result.Append($"{leftBorder}{sentence}{rightBorder}{newLine}");
+            }
+
+            // bottom
```
Comment (Almantask) [inline Src/BootCamp.Chapter/TextTable.cs]: move to a function instead of writing a comment.

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 14 above)
Comment (ygnacio) [reply Src/BootCamp.Chapter/BalanceStats.cs]: that was my intention ;-)

File: Src/BootCamp.Chapter/BalanceStats.cs (hunk 16 above)
Comment (ygnacio) [reply Src/BootCamp.Chapter/BalanceStats.cs]: I totally forgot about method overload!

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Src/BootCamp.Chapter/BalanceStats.cs  6
- Src/BootCamp.Chapter/TextTable.cs  2
- Src/BootCamp.Chapter/Program.cs  1
         

Clones added
============
- Src/BootCamp.Chapter/BalanceStats.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=5069037&bid=16678002)

**Comment by ygnacio** [comment]
Hi, do you review the last commit? anything else?

**Pull request closed without merging**